drv/sidecar-front-io/src/transceivers.rs

```diff
@@ -300,44 +301,148 @@ impl Transceivers {
+    /// LpMode signal to the module.
+    pub fn assert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitSet, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Clear LpMode bits per the specified `mask`. This directly controls the
+    /// LpMode signal to the module.
+    pub fn deassert_lpmode<M: Into<FpgaPortMasks>>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.masked_port_op(WriteOp::BitClear, mask, Addr::QSFP_MOD_LPMODE0)
+    }
+
+    /// Have the SP execute a reset by manually writing the ResetL bits in the
+    /// FPGA. SFF-8679 states that the reset pulse needs to be 10 microseconds
+    /// (t_reset_init), but the module will not be ready for normal operation
+    /// until 2 seconds (t_reset) after reset is released.
+    pub fn module_reset<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.assert_reset(mask)?;
+        userlib::hl::sleep_for(1);
+        self.deassert_reset(mask)?;
+        userlib::hl::sleep_for(2000);
+        Ok(())
+    }
+
+    /// Sequence of actions to turn off modules per the specified `mask
+    pub fn power_mode_off<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.disable_power(mask)?;
+        // lpmode is being deasserted while the module is off to
+        // because it will leak into the main power rail if
+        // actively driven in the absence of power for at least
+        // some modules.
+        self.deassert_lpmode(mask)?;
+        self.assert_reset(mask)?;
+        Ok(())
+    }
+
+    /// Sequence of actions to have modules enter low power mode per the
+    /// specified `mask
+    pub fn power_mode_low<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.enable_power(mask)?;
+        self.assert_lpmode(mask)?;
+        self.deassert_reset(mask)?;
+        Ok(())
+    }
+
+    /// Sequence of actions to have modules enter high power mode per the
+    /// specified `mask
+    pub fn power_mode_high<M: Into<FpgaPortMasks> + Copy>(
+        &self,
+        mask: M,
+    ) -> Result<(), FpgaError> {
+        self.enable_power(mask)?;
+        self.deassert_reset(mask)?;
+        self.deassert_lpmode(mask)?;
+        Ok(())
+    }
+
     /// Get the current status of all low speed signals for all ports. This is
     /// Enable, Reset, LpMode/TxDis, Power Good, Power Good Timeout, Present,
     /// and IRQ/RxLos.
     pub fn get_modules_status(&self) -> Result<ModulesStatus, FpgaError> {
-        let f0: [U16<byteorder::BigEndian>; 8] =
-            self.fpga(FpgaController::Left).read(Addr::QSFP_CTRL_EN_H)?;
-        let f1: [U16<byteorder::BigEndian>; 8] = self
+        let f0: [U16<byteorder::LittleEndian>; 8] =
```

Comment:
Was this just wrong before?

Reply:
No, since I was making some invasive changes to the FPGA's register map I standardized all the multi-byte fields to be little endian as I had been inconsistent on that before.